### PR TITLE
Fix nil upcase on startup

### PR DESCRIPTION
**Pull request by gfiorav**
## Context

A debug trace present in the HTTP header, asked for `ActionDispatch::Http::Headers#[nil]`. This results in an error trace:

```
NoMethodError (undefined method `upcase' for nil:NilClass):
  lib/carto/http_header_authentication.rb:81:in `header_value'
  lib/carto/http_header_authentication.rb:9:in `valid?'
  app/controllers/application_controller.rb:363:in `http_header_authentication?'
```

This PR provides a safety check where `ActionDispatch::Http::Headers#[possible_nil]` is only attempted if `possible_nil` is `.present?`, which means that it's neither `nil` (null) or `''` (empty string).

If the trace is still useful and must be kept (it doesn't provide much info), this check can be used to not generate the error when rendering it.

This PR also provides test to check that the code is indeed nil-resistant.

### CODE REVIEW

File: lib/carto/http_header_authentication.rb [hunk 1]
```diff
@@ -85,8 +85,7 @@ def field_from_value(request)
 
     def header_value(headers)
       header = ::Cartodb.get_config(:http_header_authentication, 'header')
-      puts "user-auto-creation : Trying to extract value from headers for #{header}, value is #{headers[header]}"
-      !header.nil? && !header.empty? ? headers[header] : nil
+      headers[header] if header.present?
```
Comment (vishalya) [inline lib/carto/http_header_authentication.rb]: We need the logs for the 'user-auto-creation', as it traces the request for the user authentication and auto-creation. the idea is that, if you search the logs for 'user-auto-creation, it should give the flow of the request - if something fails. You can either put this logging back or come up with a better solution. I agree, it looks ugly, but it serves a useful purpose.

File: spec/lib/carto/http_header_authentication_spec.rb [hunk 2]
```diff
@@ -43,6 +43,12 @@
       stub_http_header_authentication_configuration(field: 'email')
       Carto::HttpHeaderAuthentication.new.valid?(mock_email_request).should == true
     end
+
+    it 'resists when config header is nil' do
```
Comment (vishalya) [inline spec/lib/carto/http_header_authentication_spec.rb]: What's the command you run to test it? 
Also, are you able to kick off the auto-testing inside the VM? Please give me the command if you are able to do it?

**Review by vishalya — CHANGES_REQUESTED** [review]

File: spec/lib/carto/http_header_authentication_spec.rb (hunk 2 above)
Comment (gfiorav) [reply spec/lib/carto/http_header_authentication_spec.rb]: Sure!

RAILS runs tests in separate environment, `test`, that has to be migrated like the `development` or `production` environments. We provide a scrip in the `Makefile` for running all tests:

```
make check
```

This will create your testing DBs and migrate the to the latest schema before running all specs.

Manually, you'd do:
```
RAILS_ENV=test bundle exec rails db:drop
RAILS_ENV=test bundle exec rails db:create
RAILS_ENV=test bundle exec rails db:migrate
```

Now your testing environment is ready. Please be aware of the `RAILS_ENV=test`, if you don't specify it, `RAILS_ENV=development` is the default and it will wipe your db in `development`.

Now, to run any single spec (this one for exapmple), simply:
```
bundle exec rspec spec/lib/carto/http_header_authentication_spec.rb -e "resists when config header is nil"
```

Note that we don't specify the environment since the `rspec` gem sets `test` by default. You can also specify the line number where the test is at instead of its name.

File: lib/carto/http_header_authentication.rb (hunk 1 above)
Comment (gfiorav) [reply lib/carto/http_header_authentication.rb]: We do this by logging NGINX requests. We then query them with Elastic Search / Kibana.

Please be aware that you shouldn't trust these flows, since when multithreading, some messages might not make it to stdout.

Comment (gfiorav) [reply lib/carto/http_header_authentication.rb]: In the mean time, I'm putting the log back in. Do you need the values or were those for debugging? It be simpler to put `user-auto-creation : Trying to extract value from headers`, no interpolation needed.

What do you think?

Comment (vishalya) [reply lib/carto/http_header_authentication.rb]: yes, that should be good.

**Comment by gfiorav** [comment]
Hey @vishalya, I added the logging back in.

See if it's good enough :)

**Comment by vishalya** [comment]
I would like to see the values `for #{header}, value is #{headers[header]}"` if they are available... else simply hardcode to `Not available`

**Comment by gfiorav** [comment]
Ok. Btw, what do you think about the multithreading issue?

Also, we use this for logging: lib/carto/http_header_authentication.rb

In production it works with Rollbar but locally it prints to std out.

Question: what do you parse to get these messages? `log/<environment>.log`?

**Comment by vishalya** [comment]
Regarding multithreading, it's not complete. Logging is work in progress.
The way we work with logs in production is that - we collect all the logs from all the servers (nginx, rails, windshaft/sql api etc.) to kibana logstash and then we search for a key... in this auto-user-creation it will be the username and see if there are any issues.

**Comment by gfiorav** [comment]
Ok @vishalya, I think it's how you want it now

**Comment by vishalya** [comment]
Looks good, thanks, I will merge it after a round of integration testing, for the next sprint.

**Comment by gfiorav** [comment]
@vishalya since I see conflicts have appeared... is this still on course to being merged?

**Comment by vishalya** [comment]
I think, Mehmet might have merged it, if it was a part of any other PR, Let me check with him and get back to you.

**Comment by gfiorav** [comment]
Can we please merge this into blp_dev 🙏 !

**Comment by gfiorav** [comment]
Closing due to inactivity
